Add EnCodec model

hollance opened the pull request:
# What does this PR do?

Adds the EnCodec neural codec from the [High Fidelity Neural Audio Compression](https://arxiv.org/abs/2210.13438) paper.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer: @stas00, Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

hollance committed:
boilerplate stuff

hollance committed:
messing around with the feature extractor

hollance committed:
fix feature extractor

hollance committed:
unit tests for feature extractor

hollance committed:
rename speech to audio

hollance committed:
quick-and-dirty import of Meta's code

hollance committed:
import weights (sort of)

hollance committed:
cleaning up

hollance committed:
more cleaning up

hollance committed:
move encoder/decoder args into config

patrickvonplaten commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,1664 @@
+            randomly selected vector from the current batch.
+        commitment_weight (float): Weight for commitment loss.
+    """
+    def __init__(
+        self,
+        dim: int,
+        codebook_size: int,
+        codebook_dim: Optional[int] = None,
+        decay: float = 0.99,
+        epsilon: float = 1e-5,
+        kmeans_init: bool = True,
+        kmeans_iters: int = 50,
+        threshold_ema_dead_code: int = 2,
+        commitment_weight: float = 1.,
+    ):
```
```suggestion
    def __init__(self, config):
```

hollance replied:
done

patrickvonplaten commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,1664 @@
+            model += [
+                final_act(**final_activation_params)
+            ]
+        self.model = nn.Sequential(*model)
```
ok to use sequential?

patrickvonplaten replied:
Think it'd be slightly nicer / better fit transformers design to use a for loop here

hollance replied:
done

hollance committed:
cleanup model

hollance committed:
rename EnCodec -> Encodec

hollance committed:
RVQ parameters in config

patrickvonplaten committed:
add slow test

patrickvonplaten committed:
Merge branch 'encodec' of https://github.com/hollance/transformers into encodec

patrickvonplaten commented on tests/models/encodec/test_modeling_encodec.py:
```diff
@@ -0,0 +1,251 @@
+        self.assertTrue(input_values.shape == input_values_enc_dec.shape)
+
+        # Matthijs: This currently doesn't work - shouldn't it work?
+        self.assertTrue(torch.allclose(input_values, input_values_enc_dec, atol=1e-3))
```
@hollance shouldn't this work more or less?

ArthurZucker committed:
add lstm init and test_init

patrickvonplaten committed:
Add save & load

hollance committed:
finish EncodecModel

ArthurZucker committed:
remove decoder_input_values as they are ont used anywhere (not removed from doc yet)

ArthurZucker committed:
fix test feature extraction model name

ArthurZucker committed:
Merge branch 'encodec' of https://github.com/hollance/transformers into encodec

hollance committed:
cleaning up quantizer

patrickvonplaten committed:
Add better slow test

patrickvonplaten committed:
Merge branch 'encodec' of https://github.com/hollance/transformers into encodec

patrickvonplaten commented on src/transformers/models/encodec/feature_extraction_encodec.py:
```diff
@@ -0,0 +1,200 @@
+            if self.feature_size == 1 and example.shape[-1] != 1:
+                raise ValueError(f"Expected mono audio but example has {example.shape[-1]} channels")
+            if self.feature_size == 2 and example.shape[-1] != 2:
+                raise ValueError(f"Expected stereo audio but example has {example.shape[-1]} channels")
```
nice error - wonder if we should though instead throw a warning and just do conversion ourselves:
```py
stereo_audio = np.concatenate([mono_audio, mono_audio], axis=0)
```

patrickvonplaten replied:
Both ok to leave as is for now or to do automatic conversion. Maybe to begin with let's just leave it to be faster?

ArthurZucker replied:
Agreed!

patrickvonplaten committed:
Fix tests

ArthurZucker committed:
some fixup and cleaning

patrickvonplaten committed:
Improve further

patrickvonplaten commented on tests/models/encodec/test_modeling_encodec.py:
```diff
@@ -0,0 +1,380 @@
+
+@slow
+@require_torch
+class EncodecIntegrationTest(unittest.TestCase):
```
This should be all we need. Pretty much checks all the important use cases for our checkpoints

ArthurZucker committed:
Merge branch 'encodec' of https://github.com/hollance/transformers into encodec

hollance committed:
fix up conversion script

ArthurZucker committed:
test don't pass, _encode_fram does not work

ArthurZucker committed:
Merge branch 'encodec' of https://github.com/hollance/transformers into encodec

ArthurZucker committed:
update tests with output per encode and decode

hollance committed:
more cleanup

hollance committed:
rename _codebook

patrickvonplaten commented on src/transformers/models/encodec/__init__.py:
```diff
@@ -0,0 +1,80 @@
+from ...utils import (
+    OptionalDependencyNotAvailable,
+    _LazyModule,
+    is_speech_available,
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/encodec/__init__.py:
```diff
@@ -0,0 +1,80 @@
+        EncodecConfig,
+    )
+
+    try:
+        if not is_speech_available():
+            raise OptionalDependencyNotAvailable()
```
```suggestion
    try:
        if not is_speech_available():
            raise OptionalDependencyNotAvailable()
```

We don't need torchaudio to import the feature extractor, the feature extractor is very lightweight

patrickvonplaten commented on src/transformers/models/encodec/configuration_encodec.py:
```diff
@@ -0,0 +1,501 @@
+        activation="ELU",
+        activation_params={"alpha": 1.0},
+        norm="weight_norm",
+        norm_params={},
```
```suggestion
        norm_params={},
```
what is this? Do we need this?

patrickvonplaten commented on src/transformers/models/encodec/configuration_encodec.py:
```diff
@@ -0,0 +1,501 @@
+        dimension=128,
+        num_filters=32,
+        num_residual_layers=1,
+        ratios=[8, 5, 4, 2],
```
need better name here IMO

hollance committed:
remove old config cruft

patrickvonplaten commented on src/transformers/models/encodec/configuration_encodec.py:
```diff
@@ -0,0 +1,501 @@
+        dilation_base=2,
+        causal=True,
+        pad_mode="reflect",
+        true_skip=False,
```
weird name

patrickvonplaten commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,1233 @@
+    #
+    #   - if more than 2 frames overlap at a given point, we hope that by induction
+    #      something sensible happens.
+    assert len(frames)
```
we need to change this and all other `assert` statemens to `raise ValueError`

patrickvonplaten commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,1233 @@
+    return out / sum_weight
+
+
+def _apply_parametrization_norm(module: nn.Module, norm: str = 'none') -> nn.Module:
+    if norm not in _CONV_NORMALIZATIONS:
+        raise ValueError(f"Invalid normalization option: {norm}")
+    if norm == 'weight_norm':
+        return nn.utils.weight_norm(module)
+    elif norm == 'spectral_norm':
+        return nn.utils.spectral_norm(module)
+    else:
+        # We already check was in CONV_NORMALIZATION, so any other choice
+        # doesn't need reparametrization.
+        return module
```
If possible I'd actually remove this. E.g. if both checkpoints only use either `"Weight_norm"` or `'spectral_norm'` then let's just hardcode it for now

patrickvonplaten commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,1233 @@
+        return module
+
+
+def _get_norm_module(module: nn.Module, causal: bool = False, norm: str = 'none', **norm_kwargs) -> nn.Module:
+    """Return the proper normalization module. If causal is True, this will ensure the returned
+    module is causal, or return an error if the normalization doesn't support causal evaluation.
+    """
+    if norm not in _CONV_NORMALIZATIONS:
+        raise ValueError(f"Invalid normalization option: {norm}")
+    if norm == 'layer_norm':
+        assert isinstance(module, nn.modules.conv._ConvNd)
+        return ConvLayerNorm(module.out_channels, **norm_kwargs)
+    elif norm == 'time_group_norm':
+        if causal:
+            raise ValueError("GroupNorm doesn't support causal evaluation.")
+        assert isinstance(module, nn.modules.conv._ConvNd)
+        return nn.GroupNorm(1, module.out_channels, **norm_kwargs)
+    else:
+        return nn.Identity()
```
Let's try to get rid of this function. We probably don't need this function no? I'd directly add the correct classes in the code

hollance committed:
ratios & hop_length

patrickvonplaten commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,1233 @@
+        self.pad_mode = pad_mode
+
+    def forward(self, x):
+        B, C, T = x.shape
```
```suggestion
        batch, channel, time = x.shape
```

no single letter for shape names

patrickvonplaten commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,1233 @@
+
+
+@dataclass
+class QuantizedResult:
```
do we need this? Think we can delete no? Would maybe replace this with a nicer `ModelOutput` class

hollance committed:
use ModuleList instead of Sequential

patrickvonplaten commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,1233 @@
+        self.register_buffer("embed_avg", embed.clone())
+
+    @torch.jit.ignore
+    def init_embed_(self, data):
```
do we need this function  for inference? This looks like a functino that's only needed fro training

hollance committed:
clean up resnet block

ArthurZucker committed:
update types

ArthurZucker committed:
Merge branch 'encodec' of https://github.com/hollance/transformers into encodec

ArthurZucker committed:
update tests

ArthurZucker committed:
fixup

hollance committed:
quick cleanup

ArthurZucker committed:
fix padding

ArthurZucker committed:
Merge branch 'encodec' of https://github.com/hollance/transformers into encodec

ArthurZucker committed:
more styl,ing

hollance committed:
add patrick feedback

ArthurZucker committed:
fix copies

hollance commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,1280 @@
+
+
+class EncodecDecoder(nn.Module):
+    """Encodec decoder."""
```
I suggest we keep the comment here "SEANet decoder" since that's what it really is.

ArthurZucker committed:
fixup

hollance committed:
fix lstm

ArthurZucker committed:
fix shape issues

ArthurZucker committed:
Merge branch 'encodec' of https://github.com/hollance/transformers into encodec

hollance committed:
rename conv layers

ArthurZucker committed:
fixup

ArthurZucker committed:
fixup

ArthurZucker committed:
Merge branch 'encodec' of https://github.com/hollance/transformers into encodec

ArthurZucker committed:
fix decoding

hollance committed:
small conv refactoring

ArthurZucker committed:
Add padding logic

use padding mask

small conv refactoring

remove norm_params

simplify conv layers

rename conv layers

stuff

add batched test

update

Clean up

merge and update for padding

fix padding

fixup

hollance committed:
remove norm_params

hollance committed:
simplify conv layers

hollance committed:
rename conv layers

hollance committed:
stuff

hollance commented:
What still needs to be done at this point:

- remove `assert`s
- remove `einops` stuff: `rearrange`, `repeat`
- `QuantizedResult` -> `QuantizerOutput`
- I don't like the names in the `EncodecOutput` etc classes, so left some TODO items for renaming them
- rename arguments in config file (see TODO items in that file)
- add doc comments for all arguments in config file
- clean up `_linear_overlap_add`, the padding functions, and `_kmeans`
- improve variable names in the modeling code
- get rid of `activation_params` from config
- fix issues with padding on the 48khz model, since the last (incomplete) frame is different than with the original model
- remove all training code; I suggest we throw an exception from all `forward` methods
- doc strings in modeling code
- MDX file
- and probably more stuff

Most of these are small items.

patrickvonplaten committed:
Clean up

patrickvonplaten commented:
TODO list:

- <s> remove asserts </s>
- <s> remove einops stuff: rearrange, repeat </s>
QuantizedResult -> QuantizerOutput
I don't like the names in the EncodecOutput etc classes, so left some TODO items for renaming them
rename arguments in config file (see TODO items in that file)
add doc comments for all arguments in config file
- <s> clean up _linear_overlap_add, the padding functions, and _kmeans </s>
improve variable names in the modeling code
get rid of activation_params from config
fix issues with padding on the 48khz model, since the last (incomplete) frame is different than with the original model
- <s> remove all training code; I suggest we throw an exception from all forward methods </s>
doc strings in modeling code
MDX file
and probably more stuff
Most of these are small items.

ArthurZucker committed:
Merge branch 'encodec' of https://github.com/hollance/transformers into encodec

patrickvonplaten committed:
clean up more

patrickvonplaten committed:
clean up more

patrickvonplaten committed:
clean up more

patrickvonplaten committed:
More clean ups

ArthurZucker committed:
cleanup convolutions

ArthurZucker committed:
typo

ArthurZucker committed:
fix typos

ArthurZucker committed:
fixup

ArthurZucker committed:
build PR doc?

ArthurZucker committed:
start refactoring docstring

ArthurZucker committed:
fix don't pad when no strid and chunk

ArthurZucker committed:
update docstring

ArthurZucker committed:
update docstring

ArthurZucker committed:
nits

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_23655). All of your documentation changes will be reflected on that endpoint.

ArthurZucker committed:
update going to lunch

ArthurZucker committed:
update config and model

ArthurZucker committed:
fix broken testse (becaue of the config changes)

ArthurZucker committed:
fix scale computation

ArthurZucker committed:
fixu[

ArthurZucker committed:
only return dict if speciefied or if config returns it

ArthurZucker committed:
remove todos

ArthurZucker committed:
update defaults in config

ArthurZucker committed:
update conversion script

ArthurZucker committed:
fix doctest

ArthurZucker committed:
more docstring + fixup

ArthurZucker committed:
nits on batched_tests

ArthurZucker committed:
more nits

patrickvonplaten commented on docs/source/en/model_doc/encodec.mdx:
```diff
@@ -0,0 +1,42 @@
+    - decode
+    - encode
+    - forward
+
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/encodec/configuration_encodec.py:
```diff
@@ -0,0 +1,169 @@
+logger = logging.get_logger(__name__)
+
+ENCODEC_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "Matthijs/encodec_24khz": "https://huggingface.co/Matthijs/encodec_24khz/resolve/main/config.json",
+    "Matthijs/encodec_48khz": "https://huggingface.co/Matthijs/encodec_48khz/resolve/main/config.json",
```
```suggestion
    "facebook/encodec_24khz": "https://huggingface.co/facebook/encodec_24khz/resolve/main/config.json",
    "facebook/encodec_48khz": "https://huggingface.co/facebook/encodec_48khz/resolve/main/config.json",
```

patrickvonplaten commented on src/transformers/models/encodec/configuration_encodec.py:
```diff
@@ -0,0 +1,169 @@
+    This is the configuration class to store the configuration of an [`EncodecModel`]. It is used to instantiate a
+    Encodec model according to the specified arguments, defining the model architecture. Instantiating a configuration
+    with the defaults will yield a similar configuration to that of the
+    [Matthijs/encodec_24khz](https://huggingface.co/Matthijs/encodec_24khz) architecture.
```
```suggestion
    [facebook/encodec_24khz](https://huggingface.co/facebook/encodec_24khz) architecture.
```

patrickvonplaten commented on src/transformers/models/encodec/configuration_encodec.py:
```diff
@@ -0,0 +1,169 @@
+        hidden_size (`int`, *optional*, defaults to 128):
+            Intermediate representation dimension.
+        num_filters (`int`, *optional*, defaults to 32):
+            Base width for the model. ??? Width??
```
```suggestion
            Number of convolution kernels of first `EncodecConv1d` down sampling layer.
```

patrickvonplaten commented on src/transformers/models/encodec/configuration_encodec.py:
```diff
@@ -0,0 +1,169 @@
+        self.codebook_size = codebook_size
+        self.codebook_dim = codebook_dim if codebook_dim is not None else hidden_size
+
+        super().__init__(
+            **kwargs,
+        )
```
```suggestion
        super().__init__(**kwargs)
```

patrickvonplaten commented on src/transformers/models/encodec/feature_extraction_encodec.py:
```diff
@@ -0,0 +1,218 @@
+        padding_value: float = 0.0,
+        chunk_length: int = 48000,
+        chunk_stride: int = 47520,
+        return_attention_mask: bool = True,
```
Do we really need this? Shouldn't there always return an attention_mask?

patrickvonplaten commented on src/transformers/models/encodec/feature_extraction_encodec.py:
```diff
@@ -0,0 +1,218 @@
+        max_length: Optional[int] = None,
+        truncation: bool = False,
+        pad_to_multiple_of: Optional[int] = None,
+        return_attention_mask: Optional[bool] = None,
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,852 @@
+
+
+ENCODEC_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "Matthijs/encodec_24khz",
+    "Matthijs/encodec_48khz",
```
```suggestion
    "facebook/encodec_24khz",
    "facebook/encodec_48khz",
```

patrickvonplaten commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,852 @@
+            Discret code embeddings computed using `model.encode`.
+        audio_values (`torch.FlaotTensor` of shape `(batch_size, sequence_length)`, *optional*)
+            Decoded audio values, obtained using the decoder part of Encodec.
+
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,852 @@
+    """
+
+    audio_codes: torch.FloatTensor = None
+    audio_scales: Optional[torch.Tensor] = None
```
Why is this optional and the other one isn't? I would make none of the optional but all mandatory

ArthurZucker replied:
Both are mandatory indeed

patrickvonplaten commented on docs/source/en/model_doc/encodec.mdx:
```diff
@@ -0,0 +1,42 @@
+*We introduce a state-of-the-art real-time, high-fidelity, audio codec leveraging neural networks. It consists in a streaming encoder-decoder architecture with quantized latent space trained in an end-to-end fashion. We simplify and speed-up the training by using a single multiscale spectrogram adversary that efficiently reduces artifacts and produce high-quality samples. We introduce a novel los ...
+
+This model was contributed by [Matthijs](https://huggingface.co/Matthijs). 
+The original code can be found [here](https://github.com/facebookresearch/encodec).
```
I'd maybe add one example here

patrickvonplaten reviewed (approved):
Mostly nits from my side! Think we're or less ready for a final review

ArthurZucker committed:
Apply suggestions from code review

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

ArthurZucker reviewed (commented):
updating and adding missing tests

ArthurZucker committed:
update basxed on review

ArthurZucker committed:
Merge branch 'encodec' of https://github.com/hollance/transformers into encodec

ArthurZucker committed:
fix update

ArthurZucker committed:
updaet tests

sgugger commented on src/transformers/models/encodec/feature_extraction_encodec.py:
```diff
@@ -0,0 +1,212 @@
+                )
+        else:
+            logger.warning(
+                "It is strongly recommended to pass the ``sampling_rate`` argument to this function. "
```
```suggestion
                "It is strongly recommended to pass the `sampling_rate` argument to this function. "
```
Nit: no need for double backticks here.

sgugger commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,850 @@
+        Returns:
+            A list of frames containing the discrete encoded codes for the input audio waveform, along with rescaling
+            factors for each chunk when `normalize` is True. Each frames is a tuple `(codebook, scale)`, with
+            `codebook` of shape `[B, K, T]`, with `K` the number of codebooks, `T` frames.
```
Would be clearer to use longer variable names here:
```suggestion
            `codebook` of shape `[batch_size, num_codebooks, frames]`.
```

sgugger commented on tests/models/encodec/test_modeling_encodec.py:
```diff
@@ -0,0 +1,510 @@
+
+    decoder_dict = {"decoder_input_ids": decoder_input_ids} if decoder_input_ids is not None else {}
+
+    return {
+        **encoder_dict,
+        **decoder_dict,
+    }
```
```suggestion
    return {**encoder_dict, **decoder_dict}
```
Style nit

sgugger reviewed (commented):
Looking good! Just have a couple of nits

ArthurZucker committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
fixup

ArthurZucker commented:
Thanks for the review. Before asking for a final review:
- [ ] Finish designing the modelling integration tests.
- [ ] Finish testing all edge cases for the feature extractor.
Working on this!

hollance commented on src/transformers/models/encodec/feature_extraction_encodec.py:
```diff
@@ -0,0 +1,212 @@
+        feature_size: int = 1,
+        sampling_rate: int = 24000,
+        padding_value: float = 0.0,
+        chunk_length: int = 48000,
+        chunk_stride: int = 47520,
```
Do we want to pass these "magic numbers" to the feature extractor? I think it makes more sense to use the same arguments as in the config ("chunk length in seconds" and "overlap") and calculate the length and stride from those.

patrickvonplaten replied:
Agree!

ArthurZucker replied:
Agreed! Adding this

ArthurZucker replied:
We could also switch to what we usually have in pipeline asr, where the stride not the overlap but how much of the signal do you take on the left and right after moving using a step of `chunk_length` but let's not complicate that now

ArthurZucker committed:
add overlap and chunl_length_s

ArthurZucker committed:
cleanup feature extraction

ArthurZucker committed:
teste edge cases truncation and padding

ArthurZucker committed:
correct processor values

ArthurZucker committed:
update config encodec, nits

ArthurZucker committed:
fix tests

ArthurZucker committed:
fixup

ArthurZucker committed:
fix 24Hz test

ArthurZucker committed:
elle tests are green

ArthurZucker committed:
fix fixup

ArthurZucker commented on README_es.md:
```diff
@@ -267,7 +267,7 @@ Número actual de puntos de control: ![](https://img.shields.io/endpoint?url=htt
 1. **[AltCLIP](https://huggingface.co/docs/transformers/model_doc/altclip)** (from BAAI) released with the paper [AltCLIP: Altering the Language Encoder in CLIP for Extended Language Capabilities](https://arxiv.org/abs/2211.06679) by Chen, Zhongzhi and Liu, Guang and Zhang, Bo-Wen and Ye, Fulong and Yang, Qinghong and Wu, Ledell.
 1. **[Audio Spectrogram Transformer](https://huggingface.co/docs/transformers/model_doc/audio-spectrogram-transformer)** (from MIT) released with the paper [AST: Audio Spectrogram Transformer](https://arxiv.org/abs/2104.01778) by Yuan Gong, Yu-An Chung, James Glass.
-1. **[Autoformer](https://huggingface.co/docs/transformers/main/model_doc/autoformer)** (from Tsinghua University) released with the paper [Autoformer: Decomposition Transformers with Auto-Correlation for Long-Term Series Forecasting](https://arxiv.org/abs/2106.13008) by Haixu Wu, Jiehui Xu, Jianmin Wang, Mingsheng Long.
+1. **[Autoformer](https://huggingface.co/docs/transformers/model_doc/autoformer)** (from Tsinghua University) released with the paper [Autoformer: Decomposition Transformers with Auto-Correlation for Long-Term Series Forecasting](https://arxiv.org/abs/2106.13008) by Haixu Wu, Jiehui Xu, Jianmin Wang, Mingsheng Long.
```
```suggestion
1. **[Autoformer](https://huggingface.co/docs/transformers/main/model_doc/autoformer)** (from Tsinghua University) released with the paper [Autoformer: Decomposition Transformers with Auto-Correlation for Long-Term Series Forecasting](https://arxiv.org/abs/2106.13008) by Haixu Wu, Jiehui Xu, Jianmin Wang, Mingsheng Long.
```

ArthurZucker commented on README_es.md:
```diff
@@ -381,7 +382,7 @@ Número actual de puntos de control: ![](https://img.shields.io/endpoint?url=htt
 1. **[MobileNetV2](https://huggingface.co/docs/transformers/model_doc/mobilenet_v2)** (from Google Inc.) released with the paper [MobileNetV2: Inverted Residuals and Linear Bottlenecks](https://arxiv.org/abs/1801.04381) by Mark Sandler, Andrew Howard, Menglong Zhu, Andrey Zhmoginov, Liang-Chieh Chen.
 1. **[MobileViT](https://huggingface.co/docs/transformers/model_doc/mobilevit)** (from Apple) released with the paper [MobileViT: Light-weight, General-purpose, and Mobile-friendly Vision Transformer](https://arxiv.org/abs/2110.02178) by Sachin Mehta and Mohammad Rastegari.
-1. **[MobileViTV2](https://huggingface.co/docs/transformers/main/model_doc/mobilevitv2)** (from Apple) released with the paper [Separable Self-attention for Mobile Vision Transformers](https://arxiv.org/abs/2206.02680) by Sachin Mehta and Mohammad Rastegari.
+1. **[MobileViTV2](https://huggingface.co/docs/transformers/model_doc/mobilevitv2)** (from Apple) released with the paper [Separable Self-attention for Mobile Vision Transformers](https://arxiv.org/abs/2206.02680) by Sachin Mehta and Mohammad Rastegari.
```
```suggestion
1. **[MobileViTV2](https://huggingface.co/docs/transformers/main/model_doc/mobilevitv2)** (from Apple) released with the paper [Separable Self-attention for Mobile Vision Transformers](https://arxiv.org/abs/2206.02680) by Sachin Mehta and Mohammad Rastegari.
```

ArthurZucker commented on README_es.md:
```diff
@@ -423,7 +424,7 @@ Número actual de puntos de control: ![](https://img.shields.io/endpoint?url=htt
 1. **[Splinter](https://huggingface.co/docs/transformers/model_doc/splinter)** (from Tel Aviv University), released together with the paper [Few-Shot Question Answering by Pretraining Span Selection](https://arxiv.org/abs/2101.00438) by Ori Ram, Yuval Kirstain, Jonathan Berant, Amir Globerson, Omer Levy.
 1. **[SqueezeBERT](https://huggingface.co/docs/transformers/model_doc/squeezebert)** (from Berkeley) released with the paper [SqueezeBERT: What can computer vision teach NLP about efficient neural networks?](https://arxiv.org/abs/2006.11316) by Forrest N. Iandola, Albert E. Shaw, Ravi Krishna, and Kurt W. Keutzer.
-1. **[SwiftFormer](https://huggingface.co/docs/transformers/main/model_doc/swiftformer)** (from MBZUAI) released with the paper [SwiftFormer: Efficient Additive Attention for Transformer-based Real-time Mobile Vision Applications](https://arxiv.org/abs/2303.15446) by Abdelrahman Shaker, Muhammad Maaz, Hanoona Rasheed, Salman Khan, Ming-Hsuan Yang, Fahad Shahbaz Khan.
+1. **[SwiftFormer](https://huggingface.co/docs/transformers/model_doc/swiftformer)** (from MBZUAI) released with the paper [SwiftFormer: Efficient Additive Attention for Transformer-based Real-time Mobile Vision Applications](https://arxiv.org/abs/2303.15446) by Abdelrahman Shaker, Muhammad Maaz, Hanoona Rasheed, Salman Khan, Ming-Hsuan Yang, Fahad Shahbaz Khan.
```
```suggestion
1. **[SwiftFormer](https://huggingface.co/docs/transformers/main/model_doc/swiftformer)** (from MBZUAI) released with the paper [SwiftFormer: Efficient Additive Attention for Transformer-based Real-time Mobile Vision Applications](https://arxiv.org/abs/2303.15446) by Abdelrahman Shaker, Muhammad Maaz, Hanoona Rasheed, Salman Khan, Ming-Hsuan Yang, Fahad Shahbaz Khan.
```

ArthurZucker commented on docs/source/en/index.mdx:
```diff
@@ -177,7 +178,7 @@ The documentation is organized into five sections:
 1. **[Nyströmformer](model_doc/nystromformer)** (from the University of Wisconsin - Madison) released with the paper [Nyströmformer: A Nyström-Based Algorithm for Approximating Self-Attention](https://arxiv.org/abs/2102.03902) by Yunyang Xiong, Zhanpeng Zeng, Rudrasis Chakraborty, Mingxing Tan, Glenn Fung, Yin Li, Vikas Singh.
 1. **[OneFormer](model_doc/oneformer)** (from SHI Labs) released with the paper [OneFormer: One Transformer to Rule Universal Image Segmentation](https://arxiv.org/abs/2211.06220) by Jitesh Jain, Jiachen Li, MangTik Chiu, Ali Hassani, Nikita Orlov, Humphrey Shi.
-1. **[OpenLlama](model_doc/open-llama)** (from [s-JoL](https://huggingface.co/s-JoL)) released in [Open-Llama](https://github.com/s-JoL/Open-Llama). 
+1. **[OpenLlama](model_doc/open-llama)** (from [s-JoL](https://huggingface.co/s-JoL)) released in [Open-Llama](https://github.com/s-JoL/Open-Llama).
```
```suggestion
1. **[OpenLlama](model_doc/open-llama)** (from [s-JoL](https://huggingface.co/s-JoL)) released in [Open-Llama](https://github.com/s-JoL/Open-Llama). 
```

ArthurZucker commented on src/transformers/models/encodec/configuration_encodec.py:
````diff
@@ -0,0 +1,172 @@
+    ```python
+    >>> from transformers import EncodecModel, EncodecConfig
+
+    >>> # Initializing a "Matthijs/encodec_24khz" style configuration
````
```suggestion
    >>> # Initializing a "facebook/encodec_24khz" style configuration
```

ArthurZucker commented on src/transformers/models/encodec/configuration_encodec.py:
```diff
@@ -0,0 +1,172 @@
+    >>> # Initializing a "Matthijs/encodec_24khz" style configuration
+    >>> configuration = EncodecConfig()
+
+    >>> # Initializing a model (with random weights) from the "Matthijs/encodec_24khz" style configuration
```
```suggestion
    >>> # Initializing a model (with random weights) from the "facebook/encodec_24khz" style configuration
```

ArthurZucker commented on src/transformers/models/encodec/convert_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,345 @@
+        config.use_causal_conv = False
+        config.norm_type = "time_group_norm"
+        config.normalize = True
+        config.chunk_length_s = 1.0
```
```suggestion
        config.chunk_length_s = 1.0
        config.overlap = 0.01
```

ArthurZucker commented on src/transformers/models/encodec/convert_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,345 @@
+
+    model = EncodecModel(config)
+
+    feature_extractor = EncodecFeatureExtractor(feature_size=config.audio_channels, sampling_rate=config.sampling_rate)
```
```suggestion
    feature_extractor = EncodecFeatureExtractor(feature_size=config.audio_channels, sampling_rate=config.sampling_rate, chunk_length_s = config.chunk_length_s, overlap = config.overlap)
```

ArthurZucker commented on src/transformers/models/encodec/convert_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,345 @@
+
+    if repo_id:
+        print("Pushing to the hub...")
+        # feature_extractor.push_to_hub(repo_id)
```
```suggestion
        feature_extractor.push_to_hub(repo_id)
```

ArthurZucker commented on src/transformers/models/encodec/feature_extraction_encodec.py:
```diff
@@ -0,0 +1,230 @@
+    most of the main methods. Users should refer to this superclass for more information regarding those methods.
+
+    Instantiating a feature extractor with the defaults will yield a similar configuration to that of the
+    [Matthijs/encodec_24khz](https://huggingface.co/Matthijs/encodec_24khz) architecture.
```
```suggestion
    [facebook/encodec_24khz](https://huggingface.co/facebook/encodec_24khz) architecture.
```

ArthurZucker commented on src/transformers/models/encodec/feature_extraction_encodec.py:
```diff
@@ -0,0 +1,230 @@
+            Whether or not [`~EncodecFeatureExtractor.__call__`] should return `attention_mask`.
+    """
+
+    model_input_names = ["input_values", "attention_mask"]
```
```suggestion
    model_input_names = ["input_values", "padding_mask"]
```

ArthurZucker commented on src/transformers/models/encodec/feature_extraction_encodec.py:
```diff
@@ -0,0 +1,230 @@
+        overlap (`float`, *optional*):
+            Defines the overlap between each chunk. It is used to compute the `chunk_stride` using the following
+            formulae : `int((1.0 - self.overlap) * self.chunk_length)`.
+        return_attention_mask (`bool`, *optional*, defaults to `True`):
+            Whether or not [`~EncodecFeatureExtractor.__call__`] should return `attention_mask`.
```
```suggestion
```

ArthurZucker commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,850 @@
+
+            <Tip warning={true}>
+
+            `padding_mask` should only be passed if the corresponding processor has `config.return_attention_mask ==
+            True`. For all models whose processor has `config.return_attention_mask == False`, `attention_mask` should
+            **not** be passed to avoid degraded performance when doing batched inference. For such models
+            `input_values` should simply be padded with 0 and passed without `padding_mask`. Be aware that these models
+            also yield slightly different results depending on whether `input_values` is padded or not.
```
```suggestion
             `padding_mask` should always be passed, unless the input was truncated or not padded. This is because in order 
             to process tensors effectively, the input audio should be padded so that `input_length % stride = step` with `step 
             = chunk_length-stride`. This ensures that all chunks are of the same shape
```

ArthurZucker commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,850 @@
+        bandwidth (`float`, *optional*):
+            The target bandwidth. Must be one of `config.target_bandwidths`. If `None`, uses the smallest possible
+            bandwidth. bandwidth is represented as a thousandth of what it is, e.g. 6kbps bandwidth is represented as
+            bandwidth == 6.0
```
```suggestion
            `bandwidth == 6.0`
```

ArthurZucker commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,850 @@
+    Args:
+        input_values (`torch.FloatTensor` of shape `(batch_size, channels, sequence_length)`, *optional*):
+            Raw audio input converted to Float and padded to the approriate length in order to be encoded using chunks
+            of length self.chunk_length and a stride of `config.strid_length`.
```
```suggestion
            of length self.chunk_length and a stride of `config.chunk_stride`.
```

ArthurZucker commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,850 @@
+        >>> dataset = load_dataset("ashraq/esc50")
+        >>> audio_sample = dataset["train"]["audio"][0]["array"]
+
+        >>> model_id = "Matthijs/encodec_24khz"
```
```suggestion
        >>> model_id = "facebook/encodec_24khz"
```

ArthurZucker commented on tests/models/encodec/test_modeling_encodec.py:
```diff
@@ -0,0 +1,506 @@
+@require_torch
+class EncodecModelTest(ModelTesterMixin, PipelineTesterMixin, unittest.TestCase):
+    all_model_classes = (EncodecModel,) if is_torch_available() else ()
+    pipeline_model_mapping = (
+        {"automatic-speech-recognition": EncodecModel, "feature-extraction": EncodecModel}
+        if is_torch_available()
+        else {}
+    )
+    # use EnCodecForSpeechToText later on
```
```suggestion
```
Only the encoder should extract features, but let's not support this for now as it is not tranformers like

ArthurZucker commented on tests/models/encodec/test_modeling_encodec.py:
```diff
@@ -0,0 +1,506 @@
+            "3.0": 0.001,
+            "24.0": 0.0005,
+        }
+        # [298290, 308806, 305543, 320363, 315116, 320181, 308885, 303291, 4212]
```
```suggestion
```

ArthurZucker commented on tests/models/encodec/test_modeling_encodec.py:
```diff
@@ -0,0 +1,506 @@
+            self.assertTrue(rmse < expected_rmse)
+
+
+# TODO Add test where audio is truncated to a multiple + feature in the feature extractor
```
```suggestion
```

ArthurZucker committed:
Apply suggestions from code review

ArthurZucker committed:
revert readme changes

ArthurZucker committed:
fixup

ArthurZucker committed:
add example

ArthurZucker committed:
use facebook checkpoints

ArthurZucker committed:
fix typo

ArthurZucker committed:
no pipeline tests

ArthurZucker committed:
use slef.pad everywhere we can

hollance commented on docs/source/en/model_doc/encodec.mdx:
```diff
@@ -0,0 +1,59 @@
+
+*We introduce a state-of-the-art real-time, high-fidelity, audio codec leveraging neural networks. It consists in a streaming encoder-decoder architecture with quantized latent space trained in an end-to-end fashion. We simplify and speed-up the training by using a single multiscale spectrogram adversary that efficiently reduces artifacts and produce high-quality samples. We introduce a novel los ...
+
+This model was contributed by [Matthijs](https://huggingface.co/Matthijs). 
```
@ArthurZucker Don't forget to add your own name here! 😄

amyeroberts commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,848 @@
+        step = chunk_length - stride
+        if (input_length % stride) - step != 0:
+            raise ValueError(
+                "The input length is not properly padded for batched chunched decoding. Make sure to pad the input correctly."
```
```suggestion
                "The input length is not properly padded for batched chunked decoding. Make sure to pad the input correctly."
```

amyeroberts commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,848 @@
+        return quantized_out
+
+
+class EncodecResidualVectorQuantizer(nn.Module):
```
Why split out `EncodecResidualVectorQuantization` and `EncodecResidualVectorQuantizer`?

ArthurZucker replied:
No idea!

amyeroberts commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,848 @@
+        input_values (`torch.FloatTensor` of shape `(batch_size, channels, sequence_length)`, *optional*):
+            Raw audio input converted to Float and padded to the approriate length in order to be encoded using chunks
+            of length self.chunk_length and a stride of `config.chunk_stride`.
+        padding_mask (`torch.BoolTensor` of shape `(batch_size, sequence_length)`, *optional*):
```
Is this tensor shape right?

ArthurZucker replied:
Indeed, since we now use self.pad in the feature extractor, the channels are most probablya also accounted for in the padding mask

amyeroberts commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,848 @@
+            return EncodecDecoderOutput(audio_values)
+        return (audio_values,)
+
+    def _decode_frame(self, codes: torch.Tensor, scale: Optional[torch.Tensor] = None) -> torch.Tensor:
```
method should be defined before  the lines of code that call it

amyeroberts commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,848 @@
+            mask = padding_mask[..., offset : offset + chunk_length].bool()
+            frame = input_values[:, :, offset : offset + chunk_length]
+            encoded_frame, scale = self._encode_frame(frame, bandwidth, mask)
+            encoded_frames.append(encoded_frame)
+            scales.append(scale)
+
+        torch.cat([encoded[0] for encoded in encoded_frames], dim=-1)
```
To avoid iterating over twice

```suggestion
            encoded_frames.append(encoded_frame[0])
            scales.append(scale)

        torch.cat(encoded_frames, dim=-1)
```

amyeroberts commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,848 @@
+
+    @staticmethod
+    def _linear_overlap_add(frames: List[torch.Tensor], stride: int):
+        # Generic overlap add, with linear fade-in/fade-out, supporting complex scenario
```
Super nice comment ❤️

amyeroberts commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,848 @@
+        torch.cat([encoded[0] for encoded in encoded_frames], dim=-1)
+        encoded_frames = torch.stack(encoded_frames)
+
+        if return_dict:
+            return EncodecEncoderOutput(encoded_frames, scales)
+
+        return (encoded_frames, scales)
```
To match general library patterns

```suggestion
        if not return_dict:
            return (encoded_frames, scales)

        return EncodecEncoderOutput(encoded_frames, scales)
```

amyeroberts commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,848 @@
+        if padding_mask is not None and padding_mask.shape[-1] < audio_values.shape[-1]:
+            audio_values = audio_values[..., : padding_mask.shape[-1]]
+
+        if return_dict:
+            return EncodecDecoderOutput(audio_values)
+        return (audio_values,)
```
```suggestion
        if not return_dict:
            return (audio_values,)
        return EncodecDecoderOutput(audio_values)
```

amyeroberts commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,848 @@
+"""
+
+
+ENCODEC_START_DOCSTRING = r"""
```
Is this used / needed?

amyeroberts commented on tests/models/encodec/test_modeling_encodec.py:
```diff
@@ -0,0 +1,496 @@
+            model = model_class(config=configs_no_init)
+            for name, param in model.named_parameters():
+                uniform_init_parms = ["conv"]
+                # TODO find the correct init values for lstm (or let them be pytorch)
```
Done?

amyeroberts reviewed (commented):
Super nice! Thanks for adding this model

Just a few general comments: 
* I haven't dug too much into the forward pass logic, so might be overlooking something obvious. One thing I noticed is there's a lot of for loop logic, and concatenating lists to create tensors. This is going to be super slow in the forward pass. If there's any way to vectorize, I think it would really help when using the model. 

* As above, sorry if I've missed something obvious. Why don't we have the option to output the hidden states of the model? 

* There's quite a few tests which are skipped because the default test_modeling_common.py patterns don't fit. Equivalent tests should be written instead of just skipping.

ArthurZucker committed:
Apply suggestions from code review

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ArthurZucker reviewed (commented):
Thanks for reviewing on such a short notive! 🤗

ArthurZucker commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,848 @@
+    @staticmethod
+    def _pad1d(hidden_states: torch.Tensor, paddings: Tuple[int, int], mode: str = "zero", value: float = 0.0):
+        """Tiny wrapper around torch.nn.functional.pad, just to allow for reflect padding on small input.
+        If this is the case, we insert extra 0 padding to the right before the reflection happen.
```
Yes but `padding_total` is required afterwards ;)

ArthurZucker commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,848 @@
+            self.conv = nn.utils.weight_norm(self.conv)
+        elif config.norm_type == "time_group_norm":
+            self.norm = nn.GroupNorm(1, out_channels)
+
```
let's maybe raise that in the config instead of having this here and in the conv normal!

ArthurZucker replied:
Ok 👍🏻

ArthurZucker committed:
update based on review

ArthurZucker committed:
update

ArthurZucker committed:
update mdx

ArthurZucker committed:
fix bug and tests

ArthurZucker committed:
fixup

ArthurZucker commented on docs/source/en/model_doc/encodec.mdx:
```diff
@@ -0,0 +1,59 @@
+>>> encoder_outputs = model.encode(inputs["input_values"], inputs["padding_mask"])
+>>> audio_values = model.decode(encoder_outputs.audio_codes, encoder_outputs.audio_scales, inputs["padding_mask"])[0]
+# or the equivalent with a forward pass
+>>> audio_values = model(inputs["input_values"], inputs["padding_mask"]).audio_codes
```
Nope! Updated

ArthurZucker commented on src/transformers/models/encodec/feature_extraction_encodec.py:
```diff
@@ -0,0 +1,199 @@
+                max_length = min([array.shape[0] for array in raw_audio])
+                nb_step = int(np.floor(max_length / self.chunk_stride))
+                max_length = (nb_step - 1) * self.chunk_stride + self.chunk_length
+            elif padding:
```
Updated, since default is to pad, if trunc then pad will be False

ArthurZucker committed:
fix doctest

ArthurZucker committed:
remove comment

ArthurZucker commented on tests/models/encodec/test_modeling_encodec.py:
```diff
@@ -0,0 +1,496 @@
+        pass
+
+    # this model has no input embeddings
+    def test_model_common_attributes(self):
```
No, this tests that `set_input_embeddings` and `get_input_embeddings` are properly defined but we don't have that

ArthurZucker commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,848 @@
+                raise ValueError("You specified `audio_scales` but did not specify the `audio_code`")
+
+        if audio_scales is None and audio_codes is None:
+            audio_codes, audio_scales = self.encode(input_values, padding_mask, bandwidth, return_dict).to_tuple()
```
Rather if `return_dict=False`! Updating

ArthurZucker committed:
more nits

ArthurZucker committed:
add more coverage for `test_truncation_and_padding`

ArthurZucker committed:
fixup

ArthurZucker commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,848 @@
+
+        ehidden_statestra_padding = self._get_extra_padding_for_conv1d(
+            hidden_states, kernel_size, stride, padding_total
+        )
```
padding_total is needed afterwards

ArthurZucker committed:
add last test

ArthurZucker committed:
fix text

ArthurZucker committed:
nits

amyeroberts commented on src/transformers/models/encodec/feature_extraction_encodec.py:
```diff
@@ -0,0 +1,199 @@
+                - `False` or `'do_not_pad'` (default): No padding (i.e., can output a batch with sequences of different
+                  lengths).
+            truncation (`bool`):
+                Activates truncation to cut input sequences longer than *max_length* to *max_length*.
```
```suggestion
                Activates truncation to cut input sequences longer than `max_length` to `max_length`.
```

amyeroberts commented on src/transformers/models/encodec/feature_extraction_encodec.py:
```diff
@@ -0,0 +1,202 @@
+                values, a list of numpy arrays or a list of list of float values. The numpy array must be of shape
+                `(num_samples,)` for mono audio (`feature_size = 1`), or `(2, num_samples)` for stereo audio
+                (`feature_size = 2`).
+            padding (`bool`, `str` or [`~utils.PaddingStrategy`], *optional*, defaults to `False`):
```
```suggestion
            padding (`bool`, `str` or [`~utils.PaddingStrategy`], *optional*, defaults to `True`):
```

amyeroberts commented on src/transformers/models/encodec/feature_extraction_encodec.py:
```diff
@@ -0,0 +1,202 @@
+                "Failing to do so can result in silent errors that might be hard to debug."
+            )
+
+        if padding and truncation:
+            padding = False
```
I don't think this is a good idea, and isn't consistent with padding being the default behaviour. 

This means a user can pass in: 

```python
feature_extractor(inputs, padding='max_length', truncation=True)
```

and it will silently select `truncation`, even though both are specified. 

The way to do this is have `padding` default to `None`, raise an error if both `truncation` and `padding` are set and then default to `padding=True` e.g.:

```
class EncodecFeatureExtractor(...):
    def __call__(
        self, 
        ...,
        padding: Optional[bool] = None,
        truncation: Optional[bool] = None,
        ...,
    ):
    if truncation is not None and padding is not None and truncation and padding:
        raise ValueError("Cannot set both `truncation` and `padding`. Got {truncation=}, {padding=}")
    
    padding = padding if padding is not None else True
    truncation = truncation if truncation is not None else False
```

ArthurZucker replied:
Done!

amyeroberts reviewed (commented):
Looking good! Thanks for iterating so quickly on this :) 

There's just one main behaviour regarding the handling of `padding` and `truncation` flags, and returned outputs which I think needs to be addressed before merging, otherwise just a few nits and is good to go.

ArthurZucker committed:
Update tests/models/encodec/test_modeling_encodec.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ArthurZucker committed:
take care of the last comments

ArthurZucker committed:
typo

ArthurZucker committed:
fix test

ArthurZucker committed:
Merge branch 'encodec' of https://github.com/hollance/transformers into encodec

ArthurZucker committed:
nits

ArthurZucker committed:
fixup

ArthurZucker commented on src/transformers/models/encodec/convert_encodec_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,352 @@
+
+    original_checkpoint = torch.load(checkpoint_path)
+    recursively_load_weights(original_checkpoint, model, model_name)
+    model.save_pretrained(pytorch_dump_folder_path)
```
No, people can have their own fine-tuned/modified checkpoints and we don't import the model from the original library so the outputs are not fixed.

ArthurZucker commented:
Ok, addressed everything! feel free to merge if it's good with you @amyeroberts

amyeroberts reviewed (approved):
Beautiful - thanks again for adding this model and iterating!

ArthurZucker committed:
Update src/transformers/models/encodec/feature_extraction_encodec.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

fxmarty commented on src/transformers/models/encodec/modeling_encodec.py:
```diff
@@ -0,0 +1,808 @@
+            raise RuntimeError(f"Duration of frame ({duration}) is longer than chunk {self.config.chunk_length_s}")
+
+        scale = None
+        if self.config.normalize:
+            # if the padding is non zero
+            input_values = input_values * padding_mask
```
I am wondering why `padding_mask` is not used in case `normalize=False`?
